### ISSUE TITLE
array constructor accepts instance of dtype instead of dtype for the "dtype" kwarg

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#6708]
```
In [1]: np.array(1, np.int64(1))
Out[1]: array(1)

In [2]: np.array(1, np.float64(1))
Out[2]: array(1.0)
```
Same with `empty`, `zeros`, etc.
This is IMO a good place for a bug to hide.  Or perhaps there has always been this behavior?
Python 3.5, numpy 1.10.1.

**Comment by seberg** [comment]
True, seems weird to me as well, would be +1 for deprecation (modulo weird thiings that might crop up during testing). @anntzer want to have a look? This seems to hide in the general dtype conversion functions, allowing numpy scalars to be converted:
```
In [2]: np.dtype(np.float_(1))
Out[2]: dtype('float64')
```

**Comment by anntzer** [comment]
Ugh.  I had a look at `scalartypes.c.src` but it's a bit beyond my understanding of numpy internals.

**Comment by seberg** [comment]
Hahahaha, I would not look at that file voluntarily nor would I be evil enough to ask unsuspecting people to look at it ;). The functionality that causes this should be here:
```
numpy/core/src/multiarray/descriptor.c:PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
```
Of course it looks specifically written to do this, but maybe we can try to get rid of that....

**Comment by InessaPawson** [comment]
@seberg Should we deprecate this as part of NumPy 2.0?

**Comment by seberg** [comment]
I don't think there is a particular reason to tie it to 2.0.  This could be deprecated at any time.  However, I *do* somehwat expect downstream relies on `their_dtype.dtype` to be used for `np.dtype(their_dtype)`.
So there are some chances that you cannot fix this generally without providing an alternative for such usage.
